contracts/bots/PartialLiquidationBot.sol

```diff
@@ -0,0 +1,292 @@
+
+        uint256 assetOutEquivalent = IPriceOracleV3(params.priceOracle).convert(
+            amountIn, params.underlying, params.assetOut
+        ) * PERCENTAGE_FACTOR / liquidationDiscount;
+
+        // If the computed amount of assetOut is larger than balance, then only the max amount is bought,
+        // and the amount of underlying is adjusted proportionally
+        if (assetOutEquivalent > maxAmountOut) {
+            return (amountIn * maxAmountOut / assetOutEquivalent, maxAmountOut);
+        } else {
+            return (amountIn, assetOutEquivalent);
+        }
+    }
+
+    /// @dev Computes the amount of underlying to pay and assetOut to receive, given explicit amountOut
+    /// @dev The actual sum bought can be smaller than passed amountOut, as it is constrained by
+    ///      current assetOut balance and the maximal swappable underlying amount
+    function _getAmountsExactOut(LiquidationParams memory params, uint256 maxAmountIn, uint256 maxAmountOut)
+        internal
+        view
+        returns (uint256, uint256)
+    {
+        (,, uint16 liquidationDiscount,,) = ICreditManagerV3(params.creditManager).fees();
+
+        uint256 amountOut = Math.min(params.amountOut, maxAmountOut);
+
+        uint256 underlyingEquivalent = IPriceOracleV3(params.priceOracle).convert(
+            amountOut, params.assetOut, params.underlying
+        ) * liquidationDiscount / PERCENTAGE_FACTOR;
+
+        // If the computed amount of underlying is larger than max amount, then only the max amount is paid,
+        // and the amount of assetOut is adjusted proportionally
+        if (underlyingEquivalent > maxAmountIn) {
+            return (maxAmountIn, amountOut * maxAmountIn / underlyingEquivalent);
+        } else {
+            return (underlyingEquivalent, amountOut);
+        }
+    }
+
+    /// @dev Returns the multicall to execute in Credit Facade
+    function _getMultiCall(LiquidationParams memory params) internal view returns (MultiCall[] memory calls) {
+        calls = new MultiCall[](2 + (params.repay ? 1 : 0) + (params.cmVersion > 3_00 ? 1 : 0));
+
+        calls[0] = MultiCall({
+            target: params.creditFacade,
+            callData: abi.encodeCall(ICreditFacadeV3Multicall.enableToken, (params.underlying))
+        });
+
+        calls[1] = MultiCall({
+            target: params.creditFacade,
+            callData: abi.encodeCall(
+                ICreditFacadeV3Multicall.withdrawCollateral, (params.assetOut, params.amountOut, params.to)
+                )
+        });
+        if (params.repay) {
+            calls[2] = MultiCall({
+                target: params.creditFacade,
+                callData: abi.encodeCall(
+                    ICreditFacadeV3Multicall.decreaseDebt, (IERC20(params.underlying).balanceOf(params.creditAccount))
+                    )
+            });
+        }
+
+        if (params.cmVersion > 3_00) {
+            calls[4] = MultiCall({
```

Comment:
Why '4' is hardcoded here? At first glance, it's out of bounds (3 is max if length equals 4). What if repay is `false` and cmVersion > 3_00? It seems, it should be computed like `calls.length-1` or by formula similar in L240